CI boilerplate reduction

kwryankrattiger opened the pull request:
Reduce boilerplate in the CI yaml by allowing composition based on pipeline targets OS/arch

CC: @blue42u

blue42u commented:
@kwryankrattiger Thanks for working on this! One high-level question, based on [`ci_process_includes`](https://github.com/spack/spack/pull/34272/files#diff-49f4a785e6a8295a5cd5af6a072cde881c00aecefe40fadf40abb33d5c35f6f5R622) it seems like the `included` configurations are merged and then `mappings` are applied later. If so, this has the [quirky nuance I alluded to before](https://github.com/spack/spack/pull/32300#issuecomment-1319605192) and was hoping to avoid. Given this configuration sketch:
```yaml
---  # common/ci.yaml
ci:
  variables:
    SIZE: "small"
  mappings:
  - match: [llvm]
    runner-attributes:
      variables:
        SIZE: "medium"

---  # big-env/spack.yaml
spack:
  spec:
  - llvm targets=all
  - paraview
  ci:
    include: common
    variables:
      SIZE: "large"  # All the packages in this environment are large
```
In the generated pipeline, a `paraview` build job will be `SIZE:"large"`, but an `llvm` build job will be `SIZE:"medium"`. That doesn't make intuitive sense to me, if I set the `SIZE` in `spack.yaml` I would expect it to override the variable set in the `common` configuration, no matter "where" it was set.

If this PR doesn't already, is there anything preventing the `included` configurations from being applied in order, instead of merged?

kwryankrattiger commented:
@blue42u So the `ci_process_includes` function is no longer used in favor of adding a "ci.yaml" and "cdash.yaml" and "view.yaml" to spack. That said, you concern persists even with spacks default merging strategy.

To address the quirky behavior, I think that has more to do with how the "scope" is defined. At the top most level, you don't specify the value the must be applied to each job, you specify the "default" value that is populated to all jobs, including the service jobs and signing job.

The next scope that is applied is the job type scope. For the build jobs the `build-job-attributes` section is applied.

The last scope that is applied for build jobs is the mapping scope which will override and append to everything defined in the previous scopes.

This is all relatively consistent with how things were done before, but now configurations are allowed to span multiple files to move boilerplate out of the specific stack files.

To achieve the desired result of forcing all build jobs to set a variable you could either change the custom configuration scopes passed to the generate step to exclude the default mappings (subtle) or specify a mapping to all discovered packages that sets the `variables:SIZE:"large"` (less subtle).

---

One of the things that I am less thrilled about with the default spack merging is that `script` is going to merge using the prepend strategy which was the major difference between the `spack.config.merge_yaml` and `spack.ci._merge_attributes` routines. For `before_script` and `after_script` we probably want to use an `append` strategy and for `script` we always want to override.

I think maybe I am just reading the config stuff backwards and "source" prepending means the the lower precedence configs are processed last and therefore everything after is appended to it. I will run some tests once I get this working again.

But the non-override behavior for script is an issue. @tgamblin @scottwittenburg if either of you have ideas or comments on how this should work.

kwryankrattiger commented:
One the override behavior, I think I found the solution, `script:: []` should allow overriding values it seems, not sure if there is more needed for that, but that is what the docs seem to imply.

kwryankrattiger commented:
>  I do not believe the Spack config-merge can or should handle, e.g. cache: can be a list or a dict (interpreted as a 1-element list).

I did check this, and it is handled by checking if one of them is a list, converting the dict value to a list, and then applying the merge. This should be okay.

>  where one config needs to know where values are set in "previous" configs to even work.

I don't think there is really a way around this. The way spack configs work in general is read -> merge -> get processed by spack command. 

Even if each yaml was read and processed in order in the CI module, which is what we decided not to do by moving to the spack config merge instead of the `ci_process_includes`, what is the correct logic for overriding values at different scopes across different configuration precedence? I feel like the rules would be even more convoluted in that workflow than they are using the default spack workflow.

Having to know which configs are being used, the order of precedence they are specified in, and what values they set I think is an okay requirement since that is already a requirement when using configuration scopes in spack. The biggest difference is the CI module has scopes to make it easier to say "do this for everything" once at a high level rather than having to always specify configs at the lower level. Things like `SIZE` that get specified at a mapping level in general will just always need to be specified at a mapping level.

kwryankrattiger commented:
> It worries me that you're trying to add key-specific merging behavior here

The most recent changes are trying not to do this because your concerns were echoed by others. The scripts are the only ones that I am a little concerned about the default behavior being applied, but they will probably be fine.

blue42u commented:
Crazy half-baked idea, what if it was a list of "modifications" applied to the generated pipeline in reverse order:
```yaml
--- # common/ci.yaml
ci:
  pipeline-gen:  # TODO: Better name for this key
  - match_behavior: first
    submapping:
    - match: [llvm]
      build-job-attributes:
        +variables: {SIZE: "medium"}
  - build-job-attributes:
      +variables: {SIZE: "small"}

--- # big-env/spack.yaml
spack:
  include: common/ci.yaml
  spec:
  - llvm targets=all
  - paraview
  ci:
    pipeline-gen:
    - +variables: {SIZE: "large"}
```

After the Spack config-merge, the above two should become the unified config:
```yaml
spack:
  spec:
  - llvm targets=all
  - paraview
  ci:
    pipeline-gen:
    - +variables: {SIZE: "large"}
    - match_behavior: first
      submapping:
      - match: [llvm]
        build-job-attributes:
          +variables: {SIZE: "medium"}
    - build-job-attributes:
        +variables: {SIZE: "small"}
```

The entries below `pipeline-gen:` can then be "applied" to the pipeline in reverse order using a custom merge (e.g. the `+` merge prefix from #32300), and the final end result correctly prefers `spack.yaml` over `common/ci.yaml`.

Still a bit confusing from the reversed order (it would need to be explained as something like "value from first matching key" and "`+` prefix indicates use next matching sub-key"). IMHO it's still less confusing (and more maintainable) than 3 different "scopes" at which changes could be made.

kwryankrattiger commented:
> Crazy half-baked idea, what if it was a list of "modifications" applied to the generated pipeline in reverse order: 

From what I can tell, this is exactly like what is already there, but less verbose. The biggest change this is making I think is the contents of the mappings list can either be just a runner attributes dict, or a submapping/matches specification. I do like `submapping` over `matches`.

I like the idea of having a better shorthand for saying "everything gets this mapping if it gets here" that is not the somewhat odd `match: ["@:"]` hack to catch all of the things.

blue42u commented:
I wanted to keep the `build-job-attributes:` key to apply a modification only to the build jobs (and not `rebuild-index`, `no-specs-to-rebuild`, etc.). It is also needed to properly distinguish between `submapping:` the Spack key and `submapping:` the potential GitLab CI key.

But otherwise, exactly!

kwryankrattiger commented:
Mappings always only apply to the build jobs, so maybe it would make sense to nest the mappings under that section? If we go this route, the idea of having a `build-job-attributes` section kind of becomes redundant. It would make sense to me to just unify all of the job attribute mapping to the `_find_matching_config` routine and call it done. Then I could remove the extra step of merging the `build-job-attributes` section into the top level configs when initializing build job configs.

blue42u commented:
There's also `signing-job-attributes:`, and from #32300 `noop-job-attributes:`, `reindex-job-attributes:` and `pipeline-attributes:`. I would like to give them different attributes so having separate keys is important in `pipeline-gen:` entries.

The problem with `submapping` under `build-job-attribute:` is that it conflicts with the potential GitLab key `submapping:`, while if you remove `build-job-attributes:` from the `match:` clause you conflict with the potential GitLab key `match:`. It's redundant but necessary.

Basically, for #32300 the parts of the config that are GitLab CI attributes should be considered (practically) a opaque blob of data (except in `_merge_attributes`):
```yaml
--- # common/ci.yaml
ci:
  pipeline-gen:  # TODO: Better name for this key
  - match_behavior: first
    submapping:
    - match: [llvm]
      build-job-attributes: GITLAB BITS
  - build-job-attributes: GITLAB BITS

--- # big-env/spack.yaml
spack:
  include: common/ci.yaml
  spec:
  - llvm targets=all
  - paraview
  ci:
    pipeline-gen:
    - build-job-attributes: GITLAB BITS
```
So, I'm pretty sure all the `build-job-attributes:` keys are needed to make this work properly.

(It looks like there was a missing `build-job-attributes:` in my first example, sorry. This one is correct.)

kwryankrattiger commented:
Okay, I think I am seeing more where you are going with this, your idea is taking this quite a bit further than I originally was thinking, but I think that is a good thing. It certainly leans more into your work with handling then merge semantics, which I think is good.

I like this idea of the `pipeline-gen` that pushes all of the configuration blocks into a list, rather than trying to track map levels and mixed scope level between configs. In the end, it makes it more consistent when trying to reason about what you actually get in the end and validating that configs are being scoped in the correct order.

This will make the schema a tad more complicated, because there will be three types of items allowed now instead of one, `global`, `<job>-attributes`, and `submapping` items. But, in the ci module code, this could be modeled by looping the `pipeline-gen` list and calling handlers based on the type of scope encountered which would likely simplify parts of the CI module quite a bit since applying configs to jobs will always be done the same, it is only the job selection that would be different.

I would like to get more feedback from @scottwittenburg on this, but I think this is a good direction.

adamjstewart reviewed (dismissed):
I like this a lot! This resolves most of my complaints about adding new pipelines. Some additional ideas (alternatives?) worth considering, either in this PR or more likely a follow-up PR:

### Platform based on directory name

I wonder if we could condition arch on pipeline location. For example, move:
```
share/spack/gitlab/cloud_pipelines/stacks/ml-linux-x86_64-cpu/spack.yaml
```
to:
```
share/spack/gitlab/cloud_pipelines/linux/x86_64/ml-cpu/spack.yaml
```
and have the arch automatically detected based on the file location.

### Matrix of platforms

For a lot of (most?) pipelines, we want to build a set of packages on most (all?) platforms. For example, for the ML stack, I'm planning on testing the following cross product:

* platform: linux, darwin
* target: x86_64, aarch64, ppc64le
* variants: +cuda, +rocm, ~cuda~rocm

This requires me to duplicate the list of packages I want to build in 18 files. If I could instead specify a matrix in `spack.yaml` and exclude known invalid combinations (no such thing as darwin ppc64le, tensorflow doesn't build on aarch64) that would make it much easier to maintain. As is, if I want to add a new package to the list, I would have to update 18 files.

See #32894 and #32893 for details on this.

P.S. Not trying to hold up this wonderful PR by suggesting more work, just wanted to put these ideas out here so people have them in mind while working on this refactor. I'm guessing these will end up being follow-up PRs in the end.

Will review again after this is rebased, the ML pipelines have changed quite a bit since this PR branch was first created and I want to make sure nothing is lost in the refactor.

blue42u commented:
Thanks @adamjstewart for pushing this PR forward!

I'd like to request one change to be made before this is merged. Currently this PR changes the `*script:` fields representation to strings containing YAML code, I would much prefer if they were instead recursive lists of strings as is supported by the GitLab CI format. This allows sharing script-fragments between parts of the configuration by using YAML anchors, which I use frequently in my `spack ci` configuration, e.g.:
```yaml
  ci:
    pipeline-gen:
    - reindex-job:
        script:
          - &init_spack
            - git clone ...
            - source ...
          - spack buildcache update-index ...
    - build-job:
        tags: [foo]
        script:
          - *init_spack
          - spack ci rebuild ...
```
I already raised this request with @kwryankrattiger over Slack DMs, we settled on an idea to reduce surprises when using this format: `*script:` keys default to override (`*script::`) unless explicitly indicated as merge (`*script+:`), all other keys merge by default (unless explicitly indicated as override). Since "merge" for lists is confusingly prepend (instead of append) and `*script:` keys are uniquely sensitive to order, this reduces surprises by making "merge" opt-in.

kwryankrattiger commented:
@blue42u 
I think lists of lists will solve the issue I had just as well as strings, and since it is already compatible with GitLab I think that is a great solution. I will add that today and rebase on master...hopefully that goes well.

@adamjstewart 
TL;DR, I think both ideas are good but I want to keep this change mostly to changes related to YAML structure changes. The restructuring in this PR right now is just a subset of the changes it makes possible going forward.

> Platform based on directory name
I think this could make sense, then the `.gitlabci.yml` script for ml pipelines could be reused for each stack. In general, it could change the setup to just be set the platform/arch/pipeline name environment variables. Since this is more of a spack CI setup chawnge vs a yaml one I think this could come after this PR.

> Matrix of platforms
I think we get this for free using the custom config scopes. The ML pipeline should be able to maintain a common `spack.yaml` and then add that to the CLI configuration scope list in the setup pipelines. I think this is also something the can be added after this since it is not changing any of the core YAML. If it isn't already there, I think it should be relatively easy to add @tgamblin ?

kwryankrattiger committed:
CI configuration boilerplate reduction and refactor

Configuration:
- New notation for list concatenation (prepend/append)
- New notation for string concatenation (prepend/append)
- Break out configuration files for: ci.yaml, cdash.yaml, view.yaml
- Spack CI section refactored to improve self-consistency and
composability
  - Scripts are now lists of lists and/or lists of strings
  - Job attributes are now listed under precedence ordered list that are
  composed/merged using Spack config merge rules.
  - "service-jobs" are identified explicitly rather than as a batch

CI:
- Consolidate common, platform, and architecture configurations for all CI stacks into composable configuration files
- Make padding consistent across all stacks (256)
- Merge all package -> runner mappings to be consistent across all
stacks

Unit Test:
- Refactor CI module unit-tests for refactor configuration

Docs:
- Add docs for new notations in configuration.rst
- Rewrite docs on CI pipelines to be consistent with refactored CI
workflow

scottwittenburg commented on lib/spack/spack/config.py:
```diff
@@ -1044,6 +1053,34 @@ def _override(string):
+    str, str : concatenate strings.
+    [str], str : append string to list. (maybe)
+
+    """
```
This docstring looks identical to the one above, probably copy-pasted and needs update?

scottwittenburg commented on lib/spack/spack/config.py:
```diff
@@ -360,6 +363,12 @@ def _process_dict_keyname_overrides(data):
+                key.prepend = True
+            elif sk.endswith("-"):
+                key = syaml.syaml_str(sk[:-1])
+                key.append = True
```
I think these new config modifiers should be explained in the PR description.  Then the content could maybe be used to update the config section of the docs.

scottwittenburg replied:
Looks like you did this already, thanks!

scottwittenburg commented on lib/spack/spack/ci.py:
```diff
@@ -595,12 +728,36 @@ def generate_gitlab_ci_yaml(
-
-    cdash_handler = CDashHandler(yaml_root.get("cdash")) if "cdash" in yaml_root else None
+    # Add config scopes to environment
+    if True:
```
Why is this in an if block if the condition is just `True`?

kwryankrattiger replied:
I thought about making it an option..but I don't know of a good use case to not do this operation so maybe I will leave that as always on until it creates an issue.

scottwittenburg commented on lib/spack/spack/ci.py:
```diff
@@ -595,12 +728,36 @@ def generate_gitlab_ci_yaml(
+        env_includes.extend(include_scopes)
+        yaml_root["include"] = env_includes
+        with open(os.path.join(env.path, "spack.yaml"), "w") as fd:
+            fd.write(syaml.dump_config(env.yaml, default_flow_style=False))
```
Is this overwriting the active environment with the merged config?  That seems weird to me, but maybe I just don't get it yet.  If this is the intent, I think it warrants a comment.

kwryankrattiger replied:
The point of this is to add the includes to the environment to make it easier for someone using the reproducer. I think it is probably sufficient to only write this out to the copied spack.yaml as we discussed earlier.

scottwittenburg commented on lib/spack/spack/test/cmd/ci.py:
```diff
@@ -229,6 +234,13 @@ def test_ci_generate_with_env(
             assert "rebuild-index" in yaml_contents
             rebuild_job = yaml_contents["rebuild-index"]
+            with capsys.disabled():
+                print("Summary...")
```
These `print` statements should be removed I think.

scottwittenburg commented on lib/spack/spack/ci.py:
```diff
@@ -546,6 +486,235 @@ def get_spec_filter_list(env, affected_pkgs):
+            },
+        ]
+
+        # Job overrides
```
So are these the bits of different jobs that the user has no control over?  I.e. we're forcing it here, regardless of what they put?

scottwittenburg replied:
I guess that's what we did before, but now, it's easier to see what's going on :+1:

scottwittenburg commented on lib/spack/spack/ci.py:
```diff
@@ -1367,7 +1517,7 @@ def generate_gitlab_ci_yaml(
 
     with open(output_file, "w") as outf:
-        outf.write(syaml.dump_config(sorted_output, default_flow_style=True))
+        outf.write(syaml.dump_config(sorted_output, default_flow_style=False))
```
Can the flow style of the output be an option provided on the command line?  

Just spitballing the option here, but maybe something like `--output-style readable|condensed`?  I feel like most of the time, this is consumed by gitlab and never seen by a human, and adding the extra size for readability might not make sense in those situations.  The option could be `readable` by default, and we could pass `condensed` from gitlab ci or anywhere else we're concerned about yaml size.

scottwittenburg reviewed (commented):
After talking with you about this many times over the past few weeks, I'm starting to have a pretty good sense of how this will change how we define pipelines, and I think it's going to be great!

For anyone who didn't talk to you about it many times, I think this PR needs a description summarizing the changes and giving some examples.

I've added some initial questions and observations, but I'll make another pass when you're ready.

scottwittenburg commented on lib/spack/spack/ci.py:
```diff
@@ -833,6 +1026,13 @@ def generate_gitlab_ci_yaml(
         else:
             broken_spec_urls = web_util.list_url(broken_specs_url)
 
+    # Make sure all reserved tags are removed for all jobs
```
I don't think this comment makes sense here.

scottwittenburg replied:
I think my comment here is still valid, isn't it?

kwryankrattiger replied:
yup, this got moved into `generate_ir`

scottwittenburg commented on lib/spack/spack/ci.py:
```diff
@@ -546,6 +486,235 @@ def get_spec_filter_list(env, affected_pkgs):
+                do_remove = remove_job_name in section
+                do_merge = merge_job_name in section
+
+                def _apply_section(attrs, section):
```
Why define this function within the loop?  Could you move it outside to be defined only once if you pass in `do_merge` and `do_remove` as parameters?

kwryankrattiger replied:
I am using it more like a lambda with a capture list, I thought it was a little more readable to just say "apply the section to the job attributes" without the extra args. If this isn't idiomatic python I can change it though, I didn't think it cost that much to define a function in a loop.

scottwittenburg replied:
I thought it was also weird to shadow the closure variable `section` with the function parameter of the same name, but whatever.  I'm not an idiomatic python person, I just thought this was weird and suspicious when I was debugging the extra `:` getting written out in the final gitlab yaml.  But it turned out to have nothing to do with the issue, so I guess it's fine the way it is.

kwryankrattiger replied:
ah, yeah, shadowing is not really good...I will at least change the name here to make it clear it is the passed parameter not the loop parameter.

scottwittenburg commented on lib/spack/docs/pipelines.rst:
```diff
@@ -354,113 +351,27 @@ arguments you can pass to ``spack ci reproduce-build`` in order to reproduce
-for the gitlab-ci section of the spack environment file, to see precisely what
-syntax is allowed there.
+.. code-block:: bash
+  cd ${concrete_environment_dir}
```
Adding a newline here fixes the doc build error for me locally.  I did find some mention of this buried in the `Directives` section of the sphinx rst docs [here](https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#directives).

scottwittenburg commented on lib/spack/docs/pipelines.rst:
```diff
@@ -354,113 +351,27 @@ arguments you can pass to ``spack ci reproduce-build`` in order to reproduce
-^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
-Note about rebuilding buildcache index
-^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+^^^
```
You have to extend these to the end of the text they surround, or else they end up rendered like this:

![broken_section](https://user-images.githubusercontent.com/6527504/213294940-78d19e3f-2a7d-4f4a-b6a2-8db784a3f828.png)

scottwittenburg commented on lib/spack/docs/pipelines.rst:
```diff
@@ -475,21 +386,33 @@ not correctly reflect the mirror's contents at the end of a pipeline.
+adding ``rebuild-index: False`` inside the ``ci`` section of your
+spack environment.
+
+Reindex jobs do not allow modifying the ``script`` attribute since it is automatically generate using the target mirror listed in the ``mirrors::mirror`` configuration.
```
```suggestion
Reindex jobs do not allow modifying the ``script`` attribute since it is automatically generated using the target mirror listed in the ``mirrors::mirror`` configuration.
```

scottwittenburg commented on lib/spack/docs/pipelines.rst:
```diff
@@ -499,47 +422,78 @@ Following is an example of this section added to a ``spack.yaml``:
 the NO-OP job in the case of an empty pipeline.  The only field for the NO-OP
 job that might be generated for you is ``script``, but that will only happen
-if you do not provide one yourself.
+if you do not provide one yourself. Notice in this example the ``script`` uses the ``::`` notation to perscribe override behavior. Without this, the ``echo`` command would have been prepended to the automatically generated script rather than replacing it.
```
```suggestion
if you do not provide one yourself. Notice in this example the ``script`` uses the ``::`` notation to prescribe override behavior. Without this, the ``echo`` command would have been prepended to the automatically generated script rather than replacing it.
```

scottwittenburg commented on lib/spack/docs/pipelines.rst:
```diff
@@ -499,47 +422,78 @@ Following is an example of this section added to a ``spack.yaml``:
+generated, mainly how the jobs for building specs should be assigned to the
+configured runners on your instance. The main section for configuring pipelines
+is ``pipeline-gen``, which is a list of job attribute sections that are merged,
+using the same rules as Spack configs (TODO:ref:`spack_configs`), from the bottom up.
```
What is the TODO about here?  Just making a link to another section of the docs, or something more?

kwryankrattiger replied:
yeah, TODO, figure out how to reference the other section in the docs.

scottwittenburg commented on lib/spack/docs/pipelines.rst:
```diff
@@ -499,47 +422,78 @@ Following is an example of this section added to a ``spack.yaml``:
+configured runners on your instance. The main section for configuring pipelines
+is ``pipeline-gen``, which is a list of job attribute sections that are merged,
+using the same rules as Spack configs (TODO:ref:`spack_configs`), from the bottom up.
+The order sections are applied is to be consistent with how spack orders scope precendense when merging lists.
```
```suggestion
The order sections are applied is to be consistent with how spack orders scope precedence when merging lists.
```

scottwittenburg commented on lib/spack/docs/pipelines.rst:
```diff
@@ -475,21 +386,33 @@ not correctly reflect the mirror's contents at the end of a pipeline.
+Signing (signing)
+^^^
+
+This job is run after all of the rebuild jobs are completed and is intended to be used to sign the package binaries built by a protected CI run. Signing jobs are generated only if a signing job ``script`` is specified and the spack CI job type is protected. Note, if an ``any-job`` section contains a script, this will not implicitly create a ``signing`` job, a signing job may only exist if it is e ...
```
Everywhere else in this document and others, it seems newlines have been used to control line length.  We should probably do the same with all your additions here.

scottwittenburg commented on lib/spack/docs/pipelines.rst:
```diff
@@ -499,47 +422,78 @@ Following is an example of this section added to a ``spack.yaml``:
+
+Each type of job may have attributes added or removed via sections in the ``pipeline-gen`` list. Job type specific attributes may be specified using the keys ``<type>-job`` to add attributes to all jobs of type ``<type>`` or ``<type>-job-remove`` to remove attributes of type ``<type>``. Each section may only contain one type of job attribute specification, ie. , ``build-job`` and ``noop-job`` may ...
+
+All of the attributes specified are forwarded to the generated CI jobs, however special treatment is applied to the attributes ``tags``, ``image``, ``variables``, ``script``, ``before_script``, and ``after_script`` as they are components recognized explicitly by the Spack CI generator. For the ``tags`` attribute, Spack will remove reserved tags (:ref:`reserved_tags`) from all jobs specified in th ...
```
The next two paragraphs following this one probably need updated to remove mention of `runner-attributes`, which I don't think is used anymore, is it?

scottwittenburg commented on lib/spack/docs/pipelines.rst:
```diff
@@ -565,46 +519,27 @@ environment directory is located within your ``--artifacts_root`` (or if not
 you, and invokes ``spack ci rebuild``.
 
-.. _staging_algorithm:
+Sections that specify scripts (``script``, ``before_script``, ``after_script``) are all read as lists of commands or lists of lists of commands. It is recommended to write scripts as lists of lists if scripts will be composes via merging. The default behavior of merging lists will remove duplicate commands and potentially apply unwanted reordering, where as merging lists of lists will preserve th ...
```
```suggestion
Sections that specify scripts (``script``, ``before_script``, ``after_script``) are all read as lists of commands or lists of lists of commands. It is recommended to write scripts as lists of lists if scripts will be composed via merging. The default behavior of merging lists will remove duplicate commands and potentially apply unwanted reordering, whereas merging lists of lists will preserve the local ordering and never removes duplicate commands. When writing commands to the CI target script, all lists are expanded and flattened into a single list.
```

scottwittenburg commented on lib/spack/docs/pipelines.rst:
```diff
@@ -565,46 +519,27 @@ environment directory is located within your ``--artifacts_root`` (or if not
-in the ``runner-attributes`` is used to populate various parts of the job
-description that will be used by Gitlab CI. Once all the jobs have been assigned
-a runner, the ``.gitlab-ci.yml`` is written to disk.
+A special case of attribute specification is the ``submapping`` section which may be used to apply job attributes to build jobs based on the package spec associated with the rebuild job. Submapping is specified as a list of spec ``match`` lists associated with ``build-job``/``build-job-remove`` sections. There are two option for ``match_behavior``, either ``first`` or ``merge`` may be specified.  ...
```
```suggestion
A special case of attribute specification is the ``submapping`` section which may be used to apply job attributes to build jobs based on the package spec associated with the rebuild job. Submapping is specified as a list of spec ``match`` lists associated with ``build-job``/``build-job-remove`` sections. There are two options for ``match_behavior``, either ``first`` or ``merge`` may be specified. In either case, the ``submapping`` list is processed from the bottom up, and then each ``match`` list is searched for a string that satisfies the check ``spec.satisfies({match_item})`` for each concrete spec.
```

scottwittenburg commented on lib/spack/docs/configuration.rst:
```diff
@@ -239,6 +239,11 @@ lower-precedence settings. Completely ignoring higher-level configuration
 options is supported with the ``::`` notation for keys (see
 :ref:`config-overrides` below).
 
+There are also special notations for string concatenation and precendense override.
```
```suggestion
There are also special notations for string concatenation and precedence override.
```

scottwittenburg commented on lib/spack/docs/configuration.rst:
```diff
@@ -279,6 +284,47 @@ command:
+
+^^^^^^^^^^^^^^^^^^^^^^^^^^
+String Concatenation
+^^^^^^^^^^^^^^^^^^^^^^^^^^
```
While it works to extend the section specifiers past the end of the text in between, I don't see it happening anywhere else, so maybe best to adhere to the convention of making the lengths match.

scottwittenburg commented on lib/spack/docs/configuration.rst:
```diff
@@ -279,6 +284,47 @@ command:
+^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Above, the user ``config.yaml`` *completely* overrides specific settings in the
+default ``config.yaml``. Sometimes, it is useful to add addition a suffix/prefix
```
```suggestion
default ``config.yaml``. Sometimes, it is useful to add a suffix/prefix
```

scottwittenburg commented on lib/spack/docs/pipelines.rst:
```diff
@@ -9,27 +9,32 @@
+can be done quickly by setting up a local Gitlab instance.
 
-#. Create a repository on your gitlab instance
+It is possible to set up pipelines on gitlab.com the builds there are limited to
```
```suggestion
It is possible to set up pipelines on gitlab.com, but the builds there are limited to
```

scottwittenburg commented on lib/spack/docs/pipelines.rst:
```diff
@@ -9,27 +9,32 @@
+
+After setting up a Gitlab instance for running CI, the basic steps for setting up a build pipeline are as follows:
+
+#. Create a repository in the Gitlab instance CI and a runner enabled.
```
```suggestion
#. Create a repository in the Gitlab instance CI and enable a runner.
```

kwryankrattiger replied:
Since enabling CI and Runners are configs in each gitlab repo, I am trying to convey that the repo they create must have those components enabled. Would it be better as `with CI and a runner enabled`?

scottwittenburg replied:
Yeah, that sounds good to me.

scottwittenburg commented on lib/spack/docs/pipelines.rst:
```diff
@@ -132,24 +130,21 @@ And here's the spack environment built by the pipeline represented as a
+           image:
+             name: ghcr.io/scottwittenburg/ecpe4s-ubuntu18.04-runner-x86_64:2020-09-01
+             entrypoint: [""]
+
```
Does it make sense below here, to mention why you don't see any `script` section, and that one from a config scope will be inserted for you by default?

scottwittenburg commented on lib/spack/docs/pipelines.rst:
```diff
@@ -354,113 +351,27 @@ arguments you can pass to ``spack ci reproduce-build`` in order to reproduce
+^^^
 
-.. code-block:: yaml
+Rebuild jobs, denoted as ``build-job``'s in the ``pipeline-gen`` list, are jobs associated with a concrete spec that have been marked for rebuild. By default a simple script for doing rebuild is generated, but may be modified as needed.
```
```suggestion
Rebuild jobs, denoted as ``build-job``'s in the ``pipeline-gen`` list, are jobs associated with concrete specs that have been marked for rebuild. By default a simple script for doing rebuild is generated, but may be modified as needed.
```

scottwittenburg commented on lib/spack/docs/pipelines.rst:
```diff
@@ -475,21 +386,33 @@ not correctly reflect the mirror's contents at the end of a pipeline.
+Cleanup (cleanup)
+^^^
+
+When using ``temporary-storage-url-prefix`` the cleanup job will destroy the mirror created for the parent Gitlab pipeline job. Cleanup jobs do not allow modifying the script, but do expect that the spack command is in the path and require a ``before_script`` to be specified that sources the ``setup-env.sh`` script.
```
```suggestion
When using ``temporary-storage-url-prefix`` the cleanup job will destroy the mirror created for the associated Gitlab pipeline. Cleanup jobs do not allow modifying the script, but do expect that the spack command is in the path and require a ``before_script`` to be specified that sources the ``setup-env.sh`` script.
```

scottwittenburg commented on lib/spack/spack/ci.py:
```diff
@@ -546,6 +486,245 @@ def get_spec_filter_list(env, affected_pkgs):
+                "enable-artifacts-buildcache", False
+            ),
+            "bootstrap": self.ci_config.get(
+                "enable-artifacts-buildcache", []
```
```suggestion
                "bootstrap", []
```

scottwittenburg replied:
I'm not sure, but I don't think you want to mix up `bootstrap` and `enable-artifacts-buildcache` here.

scottwittenburg commented on lib/spack/spack/ci.py:
```diff
@@ -546,6 +486,245 @@ def get_spec_filter_list(env, affected_pkgs):
+            "bootstrap": self.ci_config.get(
+                "enable-artifacts-buildcache", []
+            ),  # This is deprecated and should be removed
+            "rebuild-index": self.ci_config.get("enable-artifacts-buildcache", True),
```
Same, why `enable-artifacts-buildcache` here?

scottwittenburg commented on lib/spack/spack/ci.py:
```diff
@@ -546,6 +486,245 @@ def get_spec_filter_list(env, affected_pkgs):
+                "enable-artifacts-buildcache", []
+            ),  # This is deprecated and should be removed
+            "rebuild-index": self.ci_config.get("enable-artifacts-buildcache", True),
+            "broken-specs-url": self.ci_config.get("enable-artifacts-buildcache", None),
```
ditto `enable-artifacts-buildcache`

scottwittenburg commented on lib/spack/spack/ci.py:
```diff
@@ -546,6 +486,245 @@ def get_spec_filter_list(env, affected_pkgs):
+            ),  # This is deprecated and should be removed
+            "rebuild-index": self.ci_config.get("enable-artifacts-buildcache", True),
+            "broken-specs-url": self.ci_config.get("enable-artifacts-buildcache", None),
+            "broken-tests-packages": self.ci_config.get("enable-artifacts-buildcache", []),
```
ditto `enable-artifacts-buildcache`

scottwittenburg commented on share/spack/gitlab/cloud_pipelines/configs/config.yaml:
```diff
@@ -0,0 +1,8 @@
+  build_jobs: 32
+  concretizer: clingo
+  install_tree:
+    #root: /home/software/spack
```
Is this just commented out for testing?

scottwittenburg commented on share/spack/gitlab/cloud_pipelines/configs/linux/x86_64_v1:
```diff
@@ -0,0 +1 @@
+x86_64
```
Is there such a thing as `x86_64_v1`?  I thought it was just `x86_64`.

kwryankrattiger replied:
I think so. I was mostly just copy-pasting to cover my bases. I came up with a new strategy though since the auto detection wasn't working reliably.

scottwittenburg commented on share/spack/gitlab/cloud_pipelines/configs/linux/x86_64_v2:
```diff
@@ -0,0 +1 @@
+x86_64
```
Is this missing the `_v2`?

scottwittenburg commented on share/spack/gitlab/cloud_pipelines/configs/linux/x86_64_v3:
```diff
@@ -0,0 +1 @@
+x86_64
```
Is this missing the `_v3`?

scottwittenburg commented on share/spack/gitlab/cloud_pipelines/configs/linux/x86_64_v4:
```diff
@@ -0,0 +1 @@
+x86_64
```
Is this missing the `_v4`?

scottwittenburg commented on share/spack/gitlab/cloud_pipelines/stacks/e4s-mac/spack.yaml:
```diff
@@ -39,39 +39,15 @@ spack:
-    site: Cloud Gitlab Infrastructure
+  ci:
+    pipeline-gen:
+    - cleanup-job:
```
Is this cleanup-job configured because the e4s-mac stack is using the temporary storage feature?  If so, I think both could just be removed.

kwryankrattiger replied:
I put it in to keep it consistent, but it looks like all of these macos jobs are disabled at the .gitlab-ci.yml level. I will leave removing it to @eugeneswalker since this is an E4S stack.

scottwittenburg reviewed (changes requested):
Wow, it's great to see all that yaml getting whacked out!  This pass I focused mainly on the docs, which is looking good, thank you.  I'll make another pass when you're ready.  Thanks again @kwryankrattiger  :tada: :balloon:

scottwittenburg commented on share/spack/gitlab/cloud_pipelines/configs/ci.yaml:
```diff
@@ -0,0 +1,51 @@
+        - spack env activate --without-view .
+        - spack config add "config:install_tree:projections:${SPACK_JOB_SPEC_PKG_NAME}:'morepadding/{architecture}/{compiler.name}-{compiler.version}/{name}-{version}-{hash}'"
+        - mkdir -p ${SPACK_ARTIFACTS_ROOT}/user_data
+        - if [[ -r /mnt/key/intermediate_ci_signing_key.gpg ]]; then spack gpg trust /mnt/key/intermediate_ci_signing_key.gpg; fi
```
Since you opened this PR, some things changed in the stacks on `develop`, and it won't be easy to know what might be missing, since things got moved all around.  One thing I know is missing for sure is support for UO doing protected builds and doing signing, see e.g. [here](https://github.com/spack/spack/blob/develop/share/spack/gitlab/cloud_pipelines/stacks/e4s/spack.yaml#L253-L256).  Can you add that functionality here and maybe take a quick look to see if you can spot any other differences between the script there and here?

scottwittenburg commented on lib/spack/docs/pipelines.rst:
```diff
@@ -680,6 +675,86 @@ environment/stack file, and in that case no bootstrapping will be done (only the
+``public`` and ``protected`` are used to distinguish between runners that use public
+permissions and runners with protected permissions. The ``notary`` tag is a special tag
+that is used to indicate runners that have access to the highly protected information
+used for signing binaries using a the ``signing`` job.
```
```suggestion
used for signing binaries using the ``signing`` job.
```

scottwittenburg commented on lib/spack/spack/ci.py:
```diff
@@ -546,6 +486,245 @@ def get_spec_filter_list(env, affected_pkgs):
+                elif name == "any":
+                    # Apply section attributes too all jobs
+                    for _, job in jobs.items():
+                        _apply_section(job["attributes"], section)
```
I was expecting a case for `noop-job` here, what am I missing?

kwryankrattiger replied:
noop is handled via the fall through else

scottwittenburg commented on lib/spack/spack/ci.py:
```diff
@@ -839,6 +1041,12 @@ def generate_gitlab_ci_yaml(
+    spack_ci = SpackCI(ci_config, phases, staged_phases)
+    spack_ci_ir = spack_ci.generate_ir()
+
+    with open("spack_ci_ir.txt", "w") as f:
```
Are you just writing out the ir for debugging purposes?  Should it be optional or always produced?

scottwittenburg commented on lib/spack/spack/schema/cdash.py:
```diff
@@ -15,7 +15,8 @@
         "type": "object",
         "additionalProperties": False,
-        "required": ["build-group", "url", "project", "site"],
+        # "required": ["build-group", "url", "project", "site"],
```
Why are `url`, `project` and `site` no longer required?  And if they aren't required, we probably don't want this comment in here anymore.

kwryankrattiger replied:
This is an issue that arose because part of cdash is specified in one config scope, and the rest of the config is in the stack `spack.yaml` They are all ultimately required.

scottwittenburg commented on share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml:
```diff
@@ -7,6 +7,27 @@ default:
 ########################################
 
+# CI Platform-Arch
+.linux_x86_64:
```
Should this be `.darwin_x86_64`?

scottwittenburg commented:
Whew, this is a big change!  It's looking pretty good to me at this point, but I just added a few more one-off comments for you to consider.

When do you think it will be ready for a "rebuild everything"?

kwryankrattiger commented:
> When do you think it will be ready for a "rebuild everything"?

As soon as I get a normal passing CI I am going to run the rebuild-everything

kwryankrattiger commented:
@spackbot run pipeline

kwryankrattiger commented on lib/spack/spack/ci.py:
```diff
@@ -757,6 +942,27 @@ def generate_gitlab_ci_yaml(
     shutil.copyfile(env.manifest_path, os.path.join(concrete_env_dir, "spack.yaml"))
     shutil.copyfile(env.lock_path, os.path.join(concrete_env_dir, "spack.lock"))
 
+    with open(env.manifest_path, "r") as env_fd:
```
@alalazo @tgamblin do you guys know of a better way to do this? The goal here is to add the spack config paths to the concrete environment yaml `spack.include` section so the extra config scopes only need to be specified once in the root `.gitlab-ci.yaml` rather than having to add them every time we call spack for CI.

One issue I would like to resolve is that the includes are kind of fragile. One issue that we found was the reproducer puts SPACK_ROOT in a different path than CI. This is fixable in the reproducer, but it would be nice to be able to point those includes at something like `$spack/share/spack/gitlab/cloud_pipelines/...` similarly to how the `config` section handles paths.

Another issue is there is no programmatic or composition method available to add includes to an environment. I could make one by moving includes into a "section" schema but I am not sure if that is be best course since that could in theory create recursive includes where includes include scopes of the includes that included the include scope.

kwryankrattiger commented:
@spackbot rebuild everything

kwryankrattiger commented on lib/spack/spack/environment/environment.py:
```diff
@@ -773,7 +773,7 @@ def _read_manifest(self, f, raw_yaml=None):
         self.spec_lists[user_speclist_name] = user_specs
 
-        enable_view = config_dict(self.yaml).get("view")
+        enable_view = spack.config.get("view")
```
@tgamblin @alalazo I changed this here to match moving the `view` section to a config section. In the write manifest section it looks at the env view and writes it out. I think that is fine..but maybe this needs to be re-thought now?

kwryankrattiger commented:
So in the spirit of removing hurdles, the view config is unchanged. It seems that view configuration is deeply embedded in environment and requires more in-depth investigation to handle correctly.

kwryankrattiger commented:
@spackbot rebuild everything

kwryankrattiger commented:
@spackbot rebuild everything

kwryankrattiger committed:
Script verbose environ, dev bootstrap

kwryankrattiger commented:
@spackbot rebuild everything

kwryankrattiger commented:
Itemizing all of the failures in "rebuild everything".
@scottwittenburg 

Hash issue. May a GitHub fix gone stale? @haampie I think you have seen this before [here](https://github.com/spack/spack/pull/35161)
* https://gitlab.spack.io/spack/spack/-/jobs/5446323 libxc
* https://gitlab.spack.io/spack/spack/-/jobs/5443298 libxc
* https://gitlab.spack.io/spack/spack/-/jobs/5447261 wrf
* https://gitlab.spack.io/spack/spack/-/jobs/5446343 wrf
* https://gitlab.spack.io/spack/spack/-/jobs/5447484 wrf

Other
* https://gitlab.spack.io/spack/spack/-/jobs/5447338 - Network issue
* https://gitlab.spack.io/spack/spack/-/jobs/5447682 - paraview Log too long
* https://gitlab.spack.io/spack/spack/-/jobs/5447581 - py-tensorlfow log too long
* https://gitlab.spack.io/spack/spack/-/jobs/5447695 - py-tensorlfow log too long
* https://gitlab.spack.io/spack/spack/-/jobs/5442935 - pdt%oneapi, cannot reproduce locally

Didn't run rebuild script for some reason, no artifacts
* https://gitlab.spack.io/spack/spack/-/jobs/5447304
* https://gitlab.spack.io/spack/spack/-/jobs/5447591
* https://gitlab.spack.io/spack/spack/-/jobs/5447591

Lost runner (@zackgalbreath )
* https://gitlab.spack.io/spack/spack/-/jobs/5447465
* https://gitlab.spack.io/spack/spack/-/jobs/5447466
* https://gitlab.spack.io/spack/spack/-/jobs/5447419
* https://gitlab.spack.io/spack/spack/-/jobs/5446316
* https://gitlab.spack.io/spack/spack/-/jobs/5447716
* https://gitlab.spack.io/spack/spack/-/jobs/5443251

Deprecated zlib packages
* https://gitlab.spack.io/spack/spack/-/jobs/5443182
* https://gitlab.spack.io/spack/spack/-/jobs/5443187
* https://gitlab.spack.io/spack/spack/-/jobs/5443208

kwryankrattiger committed:
Port #35409

kwryankrattiger commented:
@spackbot run pipeline

kwryankrattiger commented:
@scottwittenburg quick, review this, CI passed!

haampie commented:
Sorry for reviewing this only so late, but regarding `+:` and `-:`, I would sort of like the opposite... so that it mimics

```python
>>> a = [1, 2, 3]
>>> a += [4]
>>> a
[1, 2, 3, 4]
```

So, `key+:` would be append.
